### ISSUE TITLE
Allow chaning ZynAddSubFX internal buffer size per instrument

### ISSUE THREAD

**Issue by unfa** [LMMS/lmms#3091]
One problem I have with Zyn is that I can now clearly hear the low resolution of it's envelopes and LFOs.

I requested a HQ mode for these in the Zyn issue tracker. if you want to see the proof and undertand what I mean, read there:
https://sourceforge.net/p/zynaddsubfx/feature-requests/83/

**A quick workaround however is to change the Zyn's internal buffer size.**
Zyn updates LFOs and ENVs once every buffer, with no interpolation. This produces clicks and "zipper noise" especially when you have a big buffer and a rapid ENV of LFO action.

I measured that LMMS's implementation of Zyn uses about 5ms of internal buffer (that's 256 samples for 48 kHz). 

Mark McCurry said this about the Zyn's internal buffer in above issue:

> the plugin version defaults to 32 samples or less

Thats' 8 times smaller, giving 8 times better envelope and LFO resolution.
When you need an oscillator to go down 2 octaves in 50 ms - that's a difference between an acceptable sweep and a really poor one. And I do this all the time with drums.
Funny is I have released 3 full-length albums with drums synthesized with this low buffer and I didn't notice.

**The problem is - there's no way to change Zyn's buffer within LMMS!**

I can still live with that, and hopefully make good sounds, but it's a limitation I'm aware of now, and it makes some sounds impossible to create.

This is one area where digital synthesizers are performing worse than analogue ones. Analogue would never produces a stairsteppy envelope/LFO signal.

I'd love to be able to set this per-instrument in the Instrument tab - because I'd only change it for drum sounds that need a smooth envelope response to create real pitch slides, not a 5-starstep approximation. If I could get a 40-step approximation, I'd be very happy, no matter the CPU usage :)

PS:

I actually realized this problem after using TAL NoizeMak3r for a while (working on a Mac where I didn't have access to Zyn). That synth makes perfect sweeps, and the LFO resolution is (by my ear) as high as the sampling rate. This along with great filters and nice envelope curves makes it sound like an analogue synth (even though I never owned one). I never understood why people praise analog gear so much - but I have one good reason now (and still 5 better reasons to keep using the digital emulations).

**Comment by unfa** [comment]
I found this:

https://github.com/LMMS/lmms/blob/48cc3bb5ef85bac8cc7c01b11dd3139c547a0677/plugins/zynaddsubfx/LocalZynAddSubFx.h#L43

Does it mean that Zyn's set tu use the same buffer internally as LMMS?

A quick auditory test seems to prove this right.

Can Zyn use internally different Buffer size than LMMS?

**Comment by zonkmachine** [comment]
> Does it mean that Zyn's set tu use the same buffer internally as LMMS?

No, it just declares that the function takes an int as argument.

> Can Zyn use internally different Buffer size than LMMS?

I swapped out `bufferSize()` [here](https://github.com/LMMS/lmms/blob/48cc3bb5ef85bac8cc7c01b11dd3139c547a0677/plugins/zynaddsubfx/RemoteZynAddSubFx.cpp#L85:L88) for both `8` and `16` and it compiled and ran just fine. I don't know about the audio quality though.

**Comment by tresf** [comment]
Related? https://github.com/LMMS/lmms/pull/395/commits/cd9709d85585e1b8ae3ea09122f18f8491fd5591

**Comment by unfa** [comment]
I tried out various buffer sizes for LMMS and with 192 ZynAddSubFX is completely glitched out. I guess it's a buffer size mismatch. mybe ZYn is using 256, falling back from 192 (becasue it can't use that buffer size perhaps?) and every frame gets truncated resulting in unusable sound.

However a buffer of 4096 should result in really bad envelopes and I can't hear that.

**Comment by fundamental** [comment]
Glancing at LocalZynAddSubFx.cpp, it looks like setBufferSize and setSampleRate are rather broken. setBufferSize in particular is likely to generated crashes if it's ever increased given LMMS's wrapper. In the version included in LMMS you need to manually serialize, destroy, reinstanciate, and deserialize any Master instances for a change in either of those parameters to be correctly applied.

**Comment by unfa** [comment]
Ok, so back to the topic:

If Zyn's buffer size is not constrained by LMMS buffer size, is it possible to control the Zyn buffer for every instance? Forcing all instaces to use a lower buffer is a waste of resources, but for drums or some special FX I'd switch to 32 samples buffer instead of the (I think) default 256.

Here's a mockup I made:

![selection_159](https://cloud.githubusercontent.com/assets/491704/19815935/71c5de92-9d45-11e6-8432-406b7f31510e.png)

**Comment by fundamental** [comment]
Given the 2.5.x series of zyn code that's pretty easy. Given what LMMS currently has integrated it should be feasible for *all* zynaddsubfx instances within LMMS to have a different buffer size than the LMMS host, but it's not feasible for individual instances to have a different buffer size or sample rate.

This is due to the 'synth' global which exists in the version of code LMMS includes. This variable was removed from global scope back around early 2014 in mainline zynaddsubfx.

**Comment by unfa** [comment]
Hmm. I wonder if it's worth the hassle to make this possible before Zyn 2.5 is ready in LMMS (I heard that once 2.5 is ready, porting 3.0 is going to be a breeze).
